recorder/src/main/cpp/scheduler.cc

```diff
@@ -43,29 +44,44 @@ struct Unlocker {
     }
 };
 
-void execute_top(Scheduler::Q& q, Lock& l) {
+void execute_top(Scheduler::Q& q, Lock& l, metrics::Timer& s_t_exec) {
     Scheduler::Ent sched_for = std::move(q.top());
     q.pop();
     logger->debug("Scheduler is now triggering {}", typeid(sched_for).name());
     {
         Unlocker ul(l);
+        auto _ = s_t_exec.time_scope();
         sched_for.second();
     }
 }
 
 bool Scheduler::poll() {
     Lock l(m);
     if (q.empty()) {
+        s_m_runout.mark();
         return false;
     }
 
-    block_for_expiry(q, l, nearest_entry_changed);
+    wait_for_expiry(q, l, nearest_entry_changed, s_t_wait);
 
-    execute_top(q, l);
+    execute_top(q, l, s_t_exec);
+    auto i = 1;
 
     while ((! q.empty()) && is_expired(q.top().first)) {
-        execute_top(q, l);
+        execute_top(q, l, s_t_exec);
+        i++;
     }
+    s_h_exec_spree_len.update(i);
```

Comment:
This tells us about potential execution delays, like if 1000 things get executed in 1 run, that means 1000th one could potentially have been delayed assuming one of the first 999 was slow etc.

Its basically to catch temporal hording (thundering herd type of problem).